File: backend/app/schemas/analytics.py
from datetime import date
from typing import List, Optional
from pydantic import BaseModel


class DAUResponse(BaseModel):
    date: date
    dau: int


class ActivityPoint(BaseModel):
    date: date
    value: int


class FinancialMetricsResponse(BaseModel):
    total_revenue: float
    paying_users: int
    total_users: int
    arpu: float
    arppu: float
    conversion_rate: float
    mrr: float
    arr: float
    active_subscriptions: int
    churn_rate: float
    forecast_ltv: float
    ltv_cac_ratio: float
    cr_first: float
    cr_repeat: float


class UTMMetric(BaseModel):
    source: str
    campaign: str
    acquisitions: int
    buyers: int
    revenue: float
    conversion_rate: float


class TrafficMetric(BaseModel):
    source: str
    acquisitions: int
    buyers: int
    conversion_rate: float
    spend: float
    cac: float


class TrafficSpendRequest(BaseModel):
    project_token: str
    source: str
    spend: float


class ProductMapping(BaseModel):
    product_id: str
    billing_type: str


class ProductMappingRequest(BaseModel):
    project_token: str
    product_id: str
    billing_type: str


class UnknownProduct(BaseModel):
    product_id: str
    payments_count: int
    last_amount: float

class CohortRetention(BaseModel):
    cohort_date: str
    cohort_size: int
    retention: dict[int, float]

class FunnelStep(BaseModel):
    step_name: str
    users_count: int
    conversion_rate: float


class FunnelStepConfig(BaseModel):
    event: str
    label: str = ""


class FunnelConfigResponse(BaseModel):
    project_token: str
    steps: List[FunnelStepConfig]


class FunnelConfigRequest(BaseModel):
    project_token: str
    steps: List[FunnelStepConfig]

File: frontend/src/types/analytics.ts
export type ActivityGranularity = 'dau' | 'wau' | 'mau'

export type PeriodPreset = 'this_month' | 'last_3_months' | 'this_year'

export type BillingType = 'monthly' | 'yearly' | 'one_time'

export interface ActivityPoint {
  date: string
  value: number
}

export interface FinancialMetrics {
  total_revenue: number
  paying_users: number
  total_users: number
  arpu: number
  arppu: number
  conversion_rate: number
  mrr: number
  arr: number
  active_subscriptions: number
  churn_rate: number
  forecast_ltv: number
  ltv_cac_ratio: number
  cr_first: number
  cr_repeat: number
}

export interface TrafficMetric {
  source: string
  acquisitions: number
  buyers: number
  conversion_rate: number
  spend: number
  cac: number
}

export interface CohortRetention {
  cohort_date: string
  cohort_size: number
  retention: Record<number, number>
}

export interface FunnelStep {
  step_name: string
  users_count: number
  conversion_rate: number
}

export interface FunnelStepConfig {
  event: string
  label: string
}

export interface ProductMapping {
  product_id: string
  billing_type: BillingType
}

export interface UnknownProduct {
  product_id: string
  payments_count: number
  last_amount: number
}
